src/dials/algorithms/merging/merge.py

```diff
@@ -512,3 +524,80 @@ def process_merged_data(params, mtz_dataset, merged, merged_anomalous, stats_sum
     if MergeJSONCollector.initiated:
         stats_summary.anomalous_amplitudes = anom_amplitudes
         MergeJSONCollector.data[mtz_dataset.wavelength] = stats_summary
+
+
+def combined_miller_set(mtz_datasets: List[MTZDataClass]) -> miller.array:
+    miller_set = miller.set(
+        crystal_symmetry=mtz_datasets[0].merged_array.crystal_symmetry(),
+        indices=mtz_datasets[0].merged_array.indices().deep_copy(),
+        anomalous_flag=False,
+    )
+    for dataset in mtz_datasets[1:]:
+        indices = dataset.merged_array.indices()
+        missing_isel = miller.match_indices(miller_set.indices(), indices).singles(1)
+        miller_set.indices().extend(indices.select(missing_isel))
+    return miller_set
+
+
+def r_free_flags_from_reference(
+    params: phil.scope_extract,
+    mtz_datasets: List[MTZDataClass],
+) -> miller.array:
+
+    mtz = iotbx.mtz.object(params.r_free_flags.reference)
+    r_free_arrays = []
+    for ma in mtz.as_miller_arrays():
+        if is_rfree_array(ma, ma.info()):
+            r_free_arrays.append(ma)
+    flag_scores = get_r_free_flags_scores(
+        miller_arrays=r_free_arrays,
+        test_flag_value=params.r_free_flags.old_test_flag_value,
+    )
+    i = flag_scores.scores.index(max(flag_scores.scores))
+    r_free_array = r_free_arrays[i]
+    logger.info(
+        f"Copying FreeR flags from {params.r_free_flags.reference} with labels={r_free_array.info().labels}"
+    )
+
+    if params.r_free_flags.extend:
+        miller_set = combined_miller_set(mtz_datasets)
+        d_max, d_min = miller_set.d_max_min()
+        missing_set = r_free_array.complete_set(
+            d_min=d_min,
+            d_max=d_max,
+        ).lone_set(r_free_array.map_to_asu())
+
+        if r_free_utils.looks_like_ccp4_flags(r_free_array):
+            flag_format = "ccp4"
+        else:
+            flag_format = "cns"
+        missing_flags = missing_set.generate_r_free_flags(
+            fraction=params.r_free_flags.fraction,
+            max_free=2000,
+            lattice_symmetry_max_delta=5.0,
+            use_lattice_symmetry=params.r_free_flags.use_lattice_symmetry,
+            format=flag_format,
+        )
+        r_free_array = r_free_array.concatenate(other=missing_flags)
+
+    return r_free_array
+
+
+def generate_r_free_flags(
+    params: phil.scope_extract,
+    mtz_datasets: List[MTZDataClass],
+) -> miller.array:
+    miller_set = combined_miller_set(mtz_datasets)
+
+    if params.r_free_flags.relative_to_complete_set:
+        miller_set = miller_set.complete_set(
+            d_min=miller_set.d_min() - params.r_free_flags.d_eps
+        )
+
+    return miller_set.generate_r_free_flags(
+        fraction=params.r_free_flags.fraction,
+        max_free=2000,
+        lattice_symmetry_max_delta=5.0,
```

Comment:
```suggestion
        lattice_symmetry_max_delta=params.r_free_flags.lattice_symmetry_max_delta
```
(this value is 5.0 anyway in the default phil scope)